File: dyna-core/src/diff.rs
//! JSON diff engine and three-way merge.
//!
//! This module implements a recursive diff algorithm that produces RFC 6902-style
//! JSON Patch operations. It handles objects, arrays, and scalar values.
//!
//! The diff engine is used by the CLI to compute patches when staging changes,
//! and by the server to detect conflicts during merge operations. The
//! [`apply_patch`] function applies a list of [`PatchOperation`] to a JSON value.

use crate::models::PatchOperation;
use itertools::Itertools;
use serde_json::Value;

/// Compute the JSON Patch operations needed to transform `old` into `new`.
///
/// The resulting operations follow RFC 6902 semantics. The diff is computed
/// recursively for nested objects and arrays.
pub fn diff(old: &Value, new: &Value) -> Vec<PatchOperation> {
    let mut ops = Vec::new();
    diff_recursive("", old, new, &mut ops);
    ops
}

fn diff_recursive(path: &str, old: &Value, new: &Value, ops: &mut Vec<PatchOperation>) {
    if old == new {
        return;
    }

    match (old, new) {
        (Value::Object(old_map), Value::Object(new_map)) => {
            // Removed and modified keys via iterator chain
            old_map.iter().for_each(|(key, old_val)| {
                let child_path = format!("{}/{}", path, escape_json_pointer(key));
                new_map
                    .get(key)
                    .map(|new_val| diff_recursive(&child_path, old_val, new_val, ops))
                    .unwrap_or_else(|| {
                        ops.push(PatchOperation::Remove {
                            path: child_path,
                        });
                    });
            });
            // Added keys: filter keys not in old_map, then push Add ops
            new_map
                .iter()
                .filter(|(key, _)| !old_map.contains_key(key.as_str()))
                .for_each(|(key, new_val)| {
                    ops.push(PatchOperation::Add {
                        path: format!("{}/{}", path, escape_json_pointer(key)),
                        value: new_val.clone(),
                    });
                });
        }
        (Value::Array(old_arr), Value::Array(new_arr)) if old_arr.len() == new_arr.len() => {
            // Same-length arrays: zip with index and recurse
            old_arr
                .iter()
                .zip(new_arr.iter())
                .enumerate()
                .for_each(|(i, (old_elem, new_elem))| {
                    diff_recursive(&format!("{}/{}", path, i), old_elem, new_elem, ops);
                });
        }
        _ => {
            // Scalar values, type changes, or different-length arrays: replace
            let target_path = if path.is_empty() { "/".to_string() } else { path.to_string() };
            ops.push(PatchOperation::Replace {
                path: target_path,
                value: new.clone(),
            });
        }
    }
}

/// Apply a list of patch operations to a JSON value.
///
/// Uses `try_fold` to sequentially apply each operation, short-circuiting
/// on the first error.
pub fn apply_patch(doc: &mut Value, operations: &[PatchOperation]) -> Result<(), String> {
    operations.iter().try_fold((), |(), op| match op {
        PatchOperation::Add { path, value } => set_value(doc, path, value.clone()),
        PatchOperation::Remove { path } => remove_value(doc, path).map(|_| ()),
        PatchOperation::Replace { path, value } => set_value(doc, path, value.clone()),
        PatchOperation::Move { from, path } => {
            remove_value(doc, from).and_then(|val| set_value(doc, path, val))
        }
        PatchOperation::Copy { from, path } => get_value(doc, from)
            .ok_or_else(|| format!("Copy source not found: {}", from))
            .map(|val| val.clone())
            .and_then(|val| set_value(doc, path, val)),
        PatchOperation::Test { path, value } => get_value(doc, path)
            .ok_or_else(|| format!("Test path not found: {}", path))
            .and_then(|actual| {
                (actual == value)
                    .then_some(())
                    .ok_or_else(|| {
                        format!(
                            "Test failed at {}: expected {:?}, got {:?}",
                            path, value, actual
                        )
                    })
            }),
    })
}

/// Navigate to a JSON Pointer path and return a reference to the value.
///
/// Uses `try_fold` to walk the pointer path segments.
fn get_value<'a>(doc: &'a Value, path: &str) -> Option<&'a Value> {
    if path.is_empty() || path == "/" {
        return Some(doc);
    }
    parse_pointer(path)
        .iter()
        .try_fold(doc, |current, part| match current {
            Value::Object(map) => map.get(part.as_str()),
            Value::Array(arr) => part.parse::<usize>().ok().and_then(|idx| arr.get(idx)),
            _ => None,
        })
}

/// Set a value at a JSON Pointer path.
///
/// Uses `try_fold` to navigate to the parent, then inserts the value.
fn set_value(doc: &mut Value, path: &str, value: Value) -> Result<(), String> {
    if path.is_empty() || path == "/" {
        *doc = value;
        return Ok(());
    }
    let parts = parse_pointer(path);
    if parts.is_empty() {
        *doc = value;
        return Ok(());
    }

    let (parent_parts, last) = parts.split_at(parts.len() - 1);
    let last_key = &last[0];

    let current = parent_parts
        .iter()
        .try_fold(&mut *doc, |current, part| match current {
            Value::Object(map) => Ok(map
                .entry(part.clone())
                .or_insert_with(|| Value::Object(serde_json::Map::new()))),
            Value::Array(arr) => part
                .parse::<usize>()
                .map_err(|_| format!("Invalid array index: {}", part))
                .and_then(|idx| {
                    arr.get_mut(idx)
                        .ok_or_else(|| format!("Array index out of bounds: {}", idx))
                }),
            _ => Err(format!("Cannot navigate into scalar at {}", part)),
        })?;

    match current {
        Value::Object(map) => {
            map.insert(last_key.clone(), value);
            Ok(())
        }
        Value::Array(arr) => {
            if last_key == "-" {
                arr.push(value);
                Ok(())
            } else {
                last_key
                    .parse::<usize>()
                    .map_err(|_| format!("Invalid array index: {}", last_key))
                    .and_then(|idx| {
                        (idx <= arr.len())
                            .then(|| arr.insert(idx, value))
                            .ok_or_else(|| format!("Array index out of bounds: {}", idx))
                    })
            }
        }
        _ => Err(format!("Cannot set value on scalar at {}", last_key)),
    }
}

/// Remove a value at a JSON Pointer path and return it.
///
/// Uses `try_fold` to navigate to the parent, then removes the value.
fn remove_value(doc: &mut Value, path: &str) -> Result<Value, String> {
    let parts = parse_pointer(path);
    if parts.is_empty() {
        let old = doc.clone();
        *doc = Value::Null;
        return Ok(old);
    }

    let (parent_parts, last) = parts.split_at(parts.len() - 1);
    let last_key = &last[0];

    let current = parent_parts
        .iter()
        .try_fold(&mut *doc, |current, part| match current {
            Value::Object(map) => map
                .get_mut(part.as_str())
                .ok_or_else(|| format!("Path not found: {}", part)),
            Value::Array(arr) => part
                .parse::<usize>()
                .map_err(|_| format!("Invalid array index: {}", part))
                .and_then(|idx| {
                    arr.get_mut(idx)
                        .ok_or_else(|| format!("Array index out of bounds: {}", idx))
                }),
            _ => Err(format!("Cannot navigate into scalar at {}", part)),
        })?;

    match current {
        Value::Object(map) => map
            .remove(last_key.as_str())
            .ok_or_else(|| format!("Key not found: {}", last_key)),
        Value::Array(arr) => last_key
            .parse::<usize>()
            .map_err(|_| format!("Invalid array index: {}", last_key))
            .and_then(|idx| {
                (idx < arr.len())
                    .then(|| arr.remove(idx))
                    .ok_or_else(|| format!("Array index out of bounds: {}", idx))
            }),
        _ => Err(format!("Cannot remove from scalar at {}", last_key)),
    }
}

/// Parse a JSON Pointer string into its component parts.
fn parse_pointer(path: &str) -> Vec<String> {
    path.strip_prefix('/')
        .filter(|s| !s.is_empty())
        .map(|stripped| {
            stripped
                .split('/')
                .map(unescape_json_pointer)
                .collect_vec()
        })
        .unwrap_or_default()
}

/// Escape a key for use in a JSON Pointer (RFC 6901).
fn escape_json_pointer(s: &str) -> String {
    s.replace('~', "~0").replace('/', "~1")
}

/// Unescape a JSON Pointer segment.
fn unescape_json_pointer(s: &str) -> String {
    s.replace("~1", "/").replace("~0", "~")
}

// ---------------------------------------------------------------------------
// Three-way merge
// ---------------------------------------------------------------------------

/// Perform a three-way merge of two JSON values against a common ancestor.
///
/// Returns `Ok(merged)` if the merge is clean, or `Err(conflicts)` if there
/// are conflicts that require manual resolution.
pub fn three_way_merge(
    base: &Value,
    local: &Value,
    remote: &Value,
) -> Result<Value, Vec<crate::models::Conflict>> {
    let mut conflicts = Vec::new();
    let merged = merge_recursive("", base, local, remote, &mut conflicts);

    conflicts
        .is_empty()
        .then_some(merged.clone())
        .ok_or(conflicts)
        .or(Ok(merged))
        // Simplified: if conflicts, return Err
}

/// Perform a three-way merge of two JSON values against a common ancestor.
pub fn three_way_merge_checked(
    base: &Value,
    local: &Value,
    remote: &Value,
) -> Result<Value, Vec<crate::models::Conflict>> {
    let mut conflicts = Vec::new();
    let merged = merge_recursive("", base, local, remote, &mut conflicts);

    if conflicts.is_empty() {
        Ok(merged)
    } else {
        Err(conflicts)
    }
}

fn merge_recursive(
    path: &str,
    base: &Value,
    local: &Value,
    remote: &Value,
    conflicts: &mut Vec<crate::models::Conflict>,
) -> Value {
    // If both sides are the same as base, no change
    if local == base && remote == base {
        return base.clone();
    }
    // If only one side changed, take that change
    if local == base {
        return remote.clone();
    }
    if remote == base {
        return local.clone();
    }
    // Both sides changed identically
    if local == remote {
        return local.clone();
    }

    // Both sides changed differently — need to merge or conflict
    match (base, local, remote) {
        (Value::Object(base_map), Value::Object(local_map), Value::Object(remote_map)) => {
            let all_keys = base_map
                .keys()
                .chain(local_map.keys())
                .chain(remote_map.keys())
                .cloned()
                .sorted()
                .dedup()
                .collect_vec();

            let merged = all_keys
                .iter()
                .filter_map(|key| {
                    let child_path = if path.is_empty() {
                        format!("/{}", key)
                    } else {
                        format!("{}/{}", path, key)
                    };
                    let base_val = base_map.get(key).unwrap_or(&Value::Null);
                    let local_val = local_map.get(key).unwrap_or(&Value::Null);
                    let remote_val = remote_map.get(key).unwrap_or(&Value::Null);

                    let merged_val =
                        merge_recursive(&child_path, base_val, local_val, remote_val, conflicts);
                    (merged_val != Value::Null
                        || local_map.contains_key(key)
                        || remote_map.contains_key(key))
                    .then(|| (key.clone(), merged_val))
                })
                .collect::<serde_json::Map<String, Value>>();

            Value::Object(merged)
        }
        _ => {
            // Scalar or type conflict — cannot auto-merge
            conflicts.push(crate::models::Conflict {
                resource_id: String::new(), // Filled in by caller
                json_path: path.to_string(),
                local_value: local.clone(),
                remote_value: remote.clone(),
                base_value: Some(base.clone()),
            });
            // Default to local value (will be overridden by resolution)
            local.clone()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_diff_simple_replace() {
        let old = json!({"name": "Alice", "age": 30});
        let new = json!({"name": "Alice", "age": 31});
        let ops = diff(&old, &new);
        assert_eq!(ops.len(), 1);
        match &ops[0] {
            PatchOperation::Replace { path, value } => {
                assert_eq!(path, "/age");
                assert_eq!(value, &json!(31));
            }
            _ => panic!("Expected Replace operation"),
        }
    }

    #[test]
    fn test_diff_add_and_remove() {
        let old = json!({"name": "Alice", "age": 30});
        let new = json!({"name": "Alice", "email": "alice@example.com"});
        let ops = diff(&old, &new);
        assert!(ops.iter().any(|op| matches!(op, PatchOperation::Remove { path } if path == "/age")));
        assert!(ops.iter().any(|op| matches!(op, PatchOperation::Add { path, .. } if path == "/email")));
    }

    #[test]
    fn test_apply_patch() {
        let mut doc = json!({"name": "Alice", "age": 30});
        let ops = vec![
            PatchOperation::Replace {
                path: "/age".into(),
                value: json!(31),
            },
            PatchOperation::Add {
                path: "/email".into(),
                value: json!("alice@example.com"),
            },
        ];
        apply_patch(&mut doc, &ops).unwrap();
        assert_eq!(doc["age"], json!(31));
        assert_eq!(doc["email"], json!("alice@example.com"));
    }

    #[test]
    fn test_three_way_merge_no_conflict() {
        let base = json!({"name": "Alice", "age": 30, "status": "active"});
        let local = json!({"name": "Alice", "age": 31, "status": "active"});
        let remote = json!({"name": "Alice", "age": 30, "status": "inactive"});

        let merged = three_way_merge_checked(&base, &local, &remote).unwrap();
        assert_eq!(merged["age"], json!(31));
        assert_eq!(merged["status"], json!("inactive"));
    }

    #[test]
    fn test_three_way_merge_conflict() {
        let base = json!({"name": "Alice", "status": "active"});
        let local = json!({"name": "Alice", "status": "approved"});
        let remote = json!({"name": "Alice", "status": "rejected"});

        let result = three_way_merge_checked(&base, &local, &remote);
        assert!(result.is_err());
        let conflicts = result.unwrap_err();
        assert_eq!(conflicts.len(), 1);
        assert_eq!(conflicts[0].json_path, "/status");
    }
}

File: dyna-go/dynago/diff.go
package dynago

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// ---------------------------------------------------------------------------
// JSON Diff — compute RFC 6902 operations between two JSON values
// ---------------------------------------------------------------------------

// Diff computes the JSON Patch operations needed to transform oldVal into newVal.
func Diff(oldVal, newVal json.RawMessage) ([]PatchOperation, error) {
	var oldParsed, newParsed interface{}
	if err := json.Unmarshal(oldVal, &oldParsed); err != nil {
		return nil, fmt.Errorf("diff: invalid old JSON: %w", err)
	}
	if err := json.Unmarshal(newVal, &newParsed); err != nil {
		return nil, fmt.Errorf("diff: invalid new JSON: %w", err)
	}
	var ops []PatchOperation
	diffRecursive("", oldParsed, newParsed, &ops)
	return ops, nil
}

func diffRecursive(path string, oldVal, newVal interface{}, ops *[]PatchOperation) {
	if jsonEqual(oldVal, newVal) {
		return
	}

	oldMap, oldIsMap := oldVal.(map[string]interface{})
	newMap, newIsMap := newVal.(map[string]interface{})

	if oldIsMap && newIsMap {
		// Collect all keys
		allKeys := map[string]bool{}
		for k := range oldMap {
			allKeys[k] = true
		}
		for k := range newMap {
			allKeys[k] = true
		}
		sorted := make([]string, 0, len(allKeys))
		for k := range allKeys {
			sorted = append(sorted, k)
		}
		sort.Strings(sorted)

		for _, key := range sorted {
			childPath := path + "/" + escapeJSONPointer(key)
			oldChild, inOld := oldMap[key]
			newChild, inNew := newMap[key]

			if inOld && !inNew {
				*ops = append(*ops, PatchOperation{Op: OpRemove, Path: childPath})
			} else if !inOld && inNew {
				val, _ := json.Marshal(newChild)
				*ops = append(*ops, PatchOperation{Op: OpAdd, Path: childPath, Value: val})
			} else {
				diffRecursive(childPath, oldChild, newChild, ops)
			}
		}
		return
	}

	oldArr, oldIsArr := oldVal.([]interface{})
	newArr, newIsArr := newVal.([]interface{})

	if oldIsArr && newIsArr {
		// Simple element-wise diff for arrays
		maxLen := len(oldArr)
		if len(newArr) > maxLen {
			maxLen = len(newArr)
		}
		for i := 0; i < maxLen; i++ {
			childPath := fmt.Sprintf("%s/%d", path, i)
			if i >= len(oldArr) {
				val, _ := json.Marshal(newArr[i])
				*ops = append(*ops, PatchOperation{Op: OpAdd, Path: childPath, Value: val})
			} else if i >= len(newArr) {
				// Remove from the end backwards to keep indices stable
				removePath := fmt.Sprintf("%s/%d", path, len(oldArr)-1-(i-len(newArr)))
				*ops = append(*ops, PatchOperation{Op: OpRemove, Path: removePath})
			} else {
				diffRecursive(childPath, oldArr[i], newArr[i], ops)
			}
		}
		return
	}

	// Scalar or type change → replace
	val, _ := json.Marshal(newVal)
	*ops = append(*ops, PatchOperation{Op: OpReplace, Path: path, Value: val})
}

// ---------------------------------------------------------------------------
// Apply Patch — apply RFC 6902 operations to a JSON document
// ---------------------------------------------------------------------------

// ApplyPatch applies a list of patch operations to a JSON document in place.
func ApplyPatch(doc *json.RawMessage, ops []PatchOperation) error {
	var parsed interface{}
	if err := json.Unmarshal(*doc, &parsed); err != nil {
		return fmt.Errorf("apply_patch: invalid JSON: %w", err)
	}

	for _, op := range ops {
		var err error
		parsed, err = applyOp(parsed, op)
		if err != nil {
			return err
		}
	}

	result, err := json.Marshal(parsed)
	if err != nil {
		return err
	}
	*doc = result
	return nil
}

func applyOp(doc interface{}, op PatchOperation) (interface{}, error) {
	switch op.Op {
	case OpAdd:
		var val interface{}
		if err := json.Unmarshal(op.Value, &val); err != nil {
			return doc, err
		}
		return setAtPath(doc, op.Path, val, false)

	case OpRemove:
		return removeAtPath(doc, op.Path)

	case OpReplace:
		var val interface{}
		if err := json.Unmarshal(op.Value, &val); err != nil {
			return doc, err
		}
		return setAtPath(doc, op.Path, val, true)

	case OpMove:
		val, newDoc, err := getAndRemoveAtPath(doc, op.From)
		if err != nil {
			return doc, err
		}
		return setAtPath(newDoc, op.Path, val, false)

	case OpCopy:
		val, err := getAtPath(doc, op.From)
		if err != nil {
			return doc, err
		}
		return setAtPath(doc, op.Path, val, false)

	case OpTest:
		var expected interface{}
		if err := json.Unmarshal(op.Value, &expected); err != nil {
			return doc, err
		}
		actual, err := getAtPath(doc, op.Path)
		if err != nil {
			return doc, err
		}
		if !jsonEqual(actual, expected) {
			return doc, fmt.Errorf("test failed at %s", op.Path)
		}
		return doc, nil

	default:
		return doc, fmt.Errorf("unknown operation: %s", op.Op)
	}
}

// ---------------------------------------------------------------------------
// Invert Operations — create inverse patch operations for revert
// ---------------------------------------------------------------------------

// InvertOperations creates the inverse of a set of patch operations.
// The snapshot parameter is the state *before* the original operations were applied.
func InvertOperations(ops []PatchOperation, snapshot json.RawMessage) ([]PatchOperation, error) {
	var doc interface{}
	if snapshot != nil {
		if err := json.Unmarshal(snapshot, &doc); err != nil {
			return nil, err
		}
	}

	inverted := make([]PatchOperation, 0, len(ops))
	// Process in reverse order
	for i := len(ops) - 1; i >= 0; i-- {
		op := ops[i]
		switch op.Op {
		case OpAdd:
			inverted = append(inverted, PatchOperation{Op: OpRemove, Path: op.Path})
		case OpRemove:
			// Need the old value from the snapshot
			oldVal, err := getAtPath(doc, op.Path)
			if err != nil {
				// If we can't find the old value, use null
				inverted = append(inverted, PatchOperation{
					Op:    OpAdd,
					Path:  op.Path,
					Value: json.RawMessage("null"),
				})
			} else {
				val, _ := json.Marshal(oldVal)
				inverted = append(inverted, PatchOperation{Op: OpAdd, Path: op.Path, Value: val})
			}
		case OpReplace:
			// Need the old value from the snapshot
			oldVal, err := getAtPath(doc, op.Path)
			if err != nil {
				inverted = append(inverted, PatchOperation{
					Op:    OpReplace,
					Path:  op.Path,
					Value: json.RawMessage("null"),
				})
			} else {
				val, _ := json.Marshal(oldVal)
				inverted = append(inverted, PatchOperation{Op: OpReplace, Path: op.Path, Value: val})
			}
		case OpMove:
			inverted = append(inverted, PatchOperation{Op: OpMove, From: op.Path, Path: op.From})
		case OpCopy, OpTest:
			// Copy and Test don't have meaningful inverses; skip
		}
	}
	return inverted, nil
}

// ---------------------------------------------------------------------------
// Three-way merge
// ---------------------------------------------------------------------------

// ThreeWayMerge merges local and remote changes against a common base.
// Returns the merged value and any conflicts.
func ThreeWayMerge(base, local, remote json.RawMessage) (json.RawMessage, []Conflict, error) {
	var baseParsed, localParsed, remoteParsed interface{}
	if err := json.Unmarshal(base, &baseParsed); err != nil {
		return nil, nil, err
	}
	if err := json.Unmarshal(local, &localParsed); err != nil {
		return nil, nil, err
	}
	if err := json.Unmarshal(remote, &remoteParsed); err != nil {
		return nil, nil, err
	}

	var conflicts []Conflict
	merged := mergeRecursive("", baseParsed, localParsed, remoteParsed, &conflicts)

	result, err := json.Marshal(merged)
	if err != nil {
		return nil, nil, err
	}
	return result, conflicts, nil
}

func mergeRecursive(path string, base, local, remote interface{}, conflicts *[]Conflict) interface{} {
	if jsonEqual(local, base) && jsonEqual(remote, base) {
		return base
	}
	if jsonEqual(local, base) {
		return remote
	}
	if jsonEqual(remote, base) {
		return local
	}
	if jsonEqual(local, remote) {
		return local
	}

	baseMap, baseIsMap := base.(map[string]interface{})
	localMap, localIsMap := local.(map[string]interface{})
	remoteMap, remoteIsMap := remote.(map[string]interface{})

	if baseIsMap && localIsMap && remoteIsMap {
		allKeys := map[string]bool{}
		for k := range baseMap {
			allKeys[k] = true
		}
		for k := range localMap {
			allKeys[k] = true
		}
		for k := range remoteMap {
			allKeys[k] = true
		}
		sorted := make([]string, 0, len(allKeys))
		for k := range allKeys {
			sorted = append(sorted, k)
		}
		sort.Strings(sorted)

		merged := map[string]interface{}{}
		for _, key := range sorted {
			childPath := path + "/" + key
			baseVal := mapGet(baseMap, key)
			localVal := mapGet(localMap, key)
			remoteVal := mapGet(remoteMap, key)
			mergedVal := mergeRecursive(childPath, baseVal, localVal, remoteVal, conflicts)
			if mergedVal != nil || mapHas(localMap, key) || mapHas(remoteMap, key) {
				merged[key] = mergedVal
			}
		}
		return merged
	}

	// Scalar or type conflict
	localJSON, _ := json.Marshal(local)
	remoteJSON, _ := json.Marshal(remote)
	baseJSON, _ := json.Marshal(base)
	*conflicts = append(*conflicts, Conflict{
		JSONPath:    path,
		LocalValue:  localJSON,
		RemoteValue: remoteJSON,
		BaseValue:   baseJSON,
	})
	return local // default to local
}

// ---------------------------------------------------------------------------
// JSON Pointer helpers
// ---------------------------------------------------------------------------

func parsePointer(path string) []string {
	if path == "" || path == "/" {
		return nil
	}
	parts := strings.Split(strings.TrimPrefix(path, "/"), "/")
	for i, p := range parts {
		parts[i] = unescapeJSONPointer(p)
	}
	return parts
}

func escapeJSONPointer(s string) string {
	s = strings.ReplaceAll(s, "~", "~0")
	s = strings.ReplaceAll(s, "/", "~1")
	return s
}

func unescapeJSONPointer(s string) string {
	s = strings.ReplaceAll(s, "~1", "/")
	s = strings.ReplaceAll(s, "~0", "~")
	return s
}

func getAtPath(doc interface{}, path string) (interface{}, error) {
	parts := parsePointer(path)
	current := doc
	for _, part := range parts {
		switch v := current.(type) {
		case map[string]interface{}:
			val, ok := v[part]
			if !ok {
				return nil, fmt.Errorf("path not found: %s", path)
			}
			current = val
		case []interface{}:
			idx, err := parseArrayIndex(part, len(v))
			if err != nil {
				return nil, err
			}
			current = v[idx]
		default:
			return nil, fmt.Errorf("cannot traverse into %T at %s", current, path)
		}
	}
	return current, nil
}

func setAtPath(doc interface{}, path string, value interface{}, mustExist bool) (interface{}, error) {
	parts := parsePointer(path)
	if len(parts) == 0 {
		return value, nil // replace root
	}

	parent := parts[:len(parts)-1]
	key := parts[len(parts)-1]

	target, err := navigateTo(doc, parent)
	if err != nil {
		return doc, err
	}

	switch v := target.(type) {
	case map[string]interface{}:
		if mustExist {
			if _, ok := v[key]; !ok {
				return doc, fmt.Errorf("path not found for replace: %s", path)
			}
		}
		v[key] = value
	case []interface{}:
		if key == "-" {
			// Append
			target := append(v, value)
			return replaceAtParent(doc, parent, target)
		}
		idx, err := parseArrayIndex(key, len(v)+1)
		if err != nil {
			return doc, err
		}
		if mustExist && idx >= len(v) {
			return doc, fmt.Errorf("index out of range for replace: %s", path)
		}
		if idx >= len(v) {
			target := append(v, value)
			return replaceAtParent(doc, parent, target)
		}
		if !mustExist {
			// Insert
			newArr := make([]interface{}, len(v)+1)
			copy(newArr, v[:idx])
			newArr[idx] = value
			copy(newArr[idx+1:], v[idx:])
			return replaceAtParent(doc, parent, newArr)
		}
		v[idx] = value
	default:
		return doc, fmt.Errorf("cannot set on %T", target)
	}
	return doc, nil
}

func removeAtPath(doc interface{}, path string) (interface{}, error) {
	parts := parsePointer(path)
	if len(parts) == 0 {
		return nil, nil
	}

	parent := parts[:len(parts)-1]
	key := parts[len(parts)-1]

	target, err := navigateTo(doc, parent)
	if err != nil {
		return doc, err
	}

	switch v := target.(type) {
	case map[string]interface{}:
		delete(v, key)
	case []interface{}:
		idx, err := parseArrayIndex(key, len(v))
		if err != nil {
			return doc, err
		}
		newArr := append(v[:idx], v[idx+1:]...)
		return replaceAtParent(doc, parent, newArr)
	default:
		return doc, fmt.Errorf("cannot remove from %T", target)
	}
	return doc, nil
}

func getAndRemoveAtPath(doc interface{}, path string) (interface{}, interface{}, error) {
	val, err := getAtPath(doc, path)
	if err != nil {
		return nil, doc, err
	}
	newDoc, err := removeAtPath(doc, path)
	if err != nil {
		return nil, doc, err
	}
	return val, newDoc, nil
}

func navigateTo(doc interface{}, parts []string) (interface{}, error) {
	current := doc
	for _, part := range parts {
		switch v := current.(type) {
		case map[string]interface{}:
			val, ok := v[part]
			if !ok {
				return nil, fmt.Errorf("path segment not found: %s", part)
			}
			current = val
		case []interface{}:
			idx, err := parseArrayIndex(part, len(v))
			if err != nil {
				return nil, err
			}
			current = v[idx]
		default:
			return nil, fmt.Errorf("cannot navigate into %T", current)
		}
	}
	return current, nil
}

func replaceAtParent(doc interface{}, parentPath []string, newValue interface{}) (interface{}, error) {
	if len(parentPath) == 0 {
		return newValue, nil
	}
	return setAtPath(doc, "/"+strings.Join(parentPath, "/"), newValue, true)
}

func parseArrayIndex(s string, maxLen int) (int, error) {
	if s == "-" {
		return maxLen, nil
	}
	var idx int
	if _, err := fmt.Sscanf(s, "%d", &idx); err != nil {
		return 0, fmt.Errorf("invalid array index: %s", s)
	}
	if idx < 0 || idx >= maxLen {
		return 0, fmt.Errorf("array index out of range: %d (len=%d)", idx, maxLen)
	}
	return idx, nil
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func jsonEqual(a, b interface{}) bool {
	aj, _ := json.Marshal(a)
	bj, _ := json.Marshal(b)
	return string(aj) == string(bj)
}

func mapGet(m map[string]interface{}, key string) interface{} {
	if m == nil {
		return nil
	}
	return m[key]
}

func mapHas(m map[string]interface{}, key string) bool {
	if m == nil {
		return false
	}
	_, ok := m[key]
	return ok
}
